### PR TITLE
BUG: Remove CudaContextManager and use cudaSetDevice

### PR DESCRIPTION
**Pull request by laurendeaumatthieu**

### CODE REVIEW

**Comment by SimonRit** [comment]
Thanks for the contribution. I think you are actually cleaning up legacy code which is no longer relevant since the introduction of  a primary context in 2015... See [Cuda 7 release notes](http://developer.download.nvidia.com/compute/cuda/7_0/Prod/doc/CUDA_Toolkit_Release_Notes.pdf).
I would suggest further cleaning up the code:
- remove the ContextManager class
- replace all calls to cuSetSetCurrent to calls to `cudasetDevice` following what is [here](https://developer.nvidia.com/blog/cuda-pro-tip-always-set-current-device-avoid-multithreading-bugs/).

Should I handle it or do you want to give it a try?

**Comment by laurendeaumatthieu** [comment]
I can give a try.
What do you mean by "replace all calls to cuSetSetCurrent to calls to cudasetDevice following what is [here](https://developer.nvidia.com/blog/cuda-pro-tip-always-set-current-device-avoid-multithreading-bugs/)" ? I have removed all the `cuCtxSetCurrent`, you would like I replace them by `CUDA_CHECK(cudaSetDevice(0));` ?

**Comment by SimonRit** [comment]
Sorry for the typo. Yes, that's what I mean and what the link suggests. It shouldn't be 0 but the number of the selected device. This selection might need to be thread safe in the constructor.

**Comment by laurendeaumatthieu** [comment]
I made the changes. I let you run the tests. Please let me know if everything is ok for you

Matthieu

**Comment by SimonRit** [comment]
Thanks looks good. The CI indicates that you didn't commit the `CMakeLists.txt` file: https://my.cdash.org/build/2487527/configure. Can you also squash the changes in one commit only?
Be careful to have a commit subject "below 72 c, ideally 50" (see [here](https://docs.itk.org/en/latest/contributing/index.html)). And it would be nice to have a bit more substance in the commit message to explain why we don't need a Cuda context anymore and why we think we need to reinitialize the device before every access (with the links I provided).

**Comment by laurendeaumatthieu** [comment]
Should be ok now

**Comment by SimonRit** [comment]
Thanks. I'm surprised that it compiles locally because there are still references to the context manager. At least the CI doesn't build https://my.cdash.org/viewBuildError.php?buildid=2488764, can you fix it?

I also think the commit body is wrong as is (since the Cuda context was not destroyed) + the list of CHANGES is not necessary so I suggest the following message for the commit body:

`Use the primary context with cudaSetDevice() introduced by Cuda 7 (https://developer.download.nvidia.com/compute/cuda/7_0/Prod/doc/CUDA_Toolkit_Release_Notes.pdf) instead of a new one. cudaSetDevice is called before every memory transfer between the CPU and GPU to be sure that the context is set for the current thread, see https://developer.nvidia.com/blog/cuda-pro-tip-always-set-current-device-avoid-multithreading-bugs/.`

File: src/itkCudaDataManager.cxx [hunk 1]
```diff
@@ -16,15 +16,13 @@
  *
  *=========================================================================*/
 
-#include "itkCudaDataManager.h"
```
Comment (SimonRit) [inline src/itkCudaDataManager.cxx:19]: Oops, erroneous deletion? https://my.cdash.org/viewBuildError.php?buildid=2488836

Comment (laurendeaumatthieu) [reply src/itkCudaDataManager.cxx]: Yes, sorry

File: src/itkCudaDataManager.cxx [hunk 2]
```diff
@@ -24,7 +24,6 @@ namespace itk
 // constructor
 CudaDataManager::CudaDataManager()
 {
-  m_ContextManager = CudaContextManager::GetInstance();
```
Comment (SimonRit) [inline src/itkCudaDataManager.cxx:27]: Can you remove this long comment too? I think it was already obsolete. Sorry, I should have caught this before...

Comment (laurendeaumatthieu) [reply src/itkCudaDataManager.cxx]: Done

**Comment by SimonRit** [comment]
Thanks. Failing python packages are being solved in a separate issue, I'm merging this one.

**Pull request merged**